### PR TITLE
feat: support configuring TLS via Helm chart [DET-4030]

### PR DESCRIPTION
**Pull request by aaron276h**
## Description
Added support for configuring TLS to terminate inside a load-balancer or inside the Determined
master. 

## Test Plan
Deployed both methods on a running k8 cluster and confirmed that TLS encryption worked 
(with certificates signed by LetsEncrypt).

### CODE REVIEW

File: docs/how-to/installation/kubernetes.txt [hunk 1]
```diff
@@ -43,13 +43,21 @@ When the Determined Helm chart is installed, the following entities will
 be created:
 
 #. Deployment of the Determined master.
+
 #. ConfigMap containing configurations for the Determined master.
-#. LoadBalancer service to make the Determined master accessible.
+
+#. LoadBalancer service to make the Determined master accessible. Later
+   in this guide, we descrive how it is possible to replace this with a
```
[the remark this answers is not in the document]
Comment (aaron276h) [reply docs/how-to/installation/kubernetes.txt]: fixed

File: helm/charts/determined/templates/master-config.yaml [hunk 2]
```diff
@@ -35,7 +35,14 @@ data:
       port: {{ .Values.db.port }}
       name: {{ .Values.db.name | quote }}
 
-    http_port: {{ .Values.httpPort }}
+    {{- if .Values.tlsSecret }}
+    security:
+      tls:
+        cert: /mount/determined/secrets/tls.crt
```
[the remark this answers is not in the document]
Comment (aaron276h) [reply helm/charts/determined/templates/master-config.yaml]: Yep there is, added it in

File: helm/charts/determined/templates/master-service.yaml [hunk 3]
```diff
@@ -8,8 +8,17 @@ metadata:
     release: {{ .Release.Name }}
 spec:
   ports:
-  - port: {{ .Values.httpPort | default 8080 }}
+  - port: {{ required "A valid Values.masterPort entry required!" .Values.masterPort }}
+    {{ $targetPort := 8080 }}
+    {{- if .Values.tlsSecret -}}
+      {{ $targetPort = 8443 }}
+    {{- end }}
+    targetPort: {{ $targetPort }}
```
[the remark this answers is not in the document]
Comment (aaron276h) [reply helm/charts/determined/templates/master-service.yaml]: So i just always set Determined master port to be 8081 and then we don't even need this if statement.

File: helm/charts/determined/templates/master-service.yaml [hunk 4]
```diff
@@ -8,8 +8,17 @@ metadata:
...
     protocol: TCP
-  type: LoadBalancer
+  {{ $serviceType := "LoadBalancer" }}
+  {{- if .Values.useNodePortForMaster | default false -}}
```
[the remark this answers is not in the document]
Comment (aaron276h) [reply helm/charts/determined/templates/master-service.yaml:15]: do you mean `Values.masterPort`? That one is a `required` field so we don't need to set a default for it.

File: helm/charts/determined/values.yaml [hunk 5]
```diff
@@ -1,5 +1,20 @@
...
+
+# When useNodePortForMaster is set to false (default), a LoadBalancer service is deployed
+# to make the Determined master reachable from outside the cluster. When useNodePortForMaster
+# is set to true, the master will instead be exposed behind a NodePort service. When using a
+# NodePort service users will typically have to configure an Ingress to make the Determined
+# master reachable from outside the cluster. NodePort service is recommended when configuring
+# TLS termination in a load-balancer.
+useNodePortForMaster: false
+
+# tlsSecret enables TLS encryption for all communication with the Determined master (TLS
```
[the remark this answers is not in the document]
Comment (aaron276h) [reply helm/charts/determined/values.yaml]: updated

File: helm/charts/determined/templates/master-service.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (aaron276h) [reply helm/charts/determined/templates/master-service.yaml]: Also i had trouble getting the inline working correctly with Helm, will look into it again but want to get this landed for now.

File: helm/charts/determined/templates/master-config.yaml [hunk 6]
```diff
@@ -35,7 +35,13 @@ data:
       port: {{ .Values.db.port }}
       name: {{ .Values.db.name | quote }}
 
-    http_port: {{ .Values.httpPort }}
+    {{- if .Values.tlsSecret }}
+    security:
+      tls:
+        cert: {{- include "determined.secretPath" . | indent 1 -}}tls.crt
```
Comment (dzhu) [inline helm/charts/determined/templates/master-config.yaml:41]: non-blocking: Would be much more natural to just drop the `-`s and not use `indent`, here and in various other places.

**Review by dzhu — APPROVED** [review]
Got another template nitpick if you feel like tweaking it again, but feel free to go ahead.

File: helm/charts/determined/templates/master-config.yaml (hunk 6 above)
Comment (dzhu) [reply helm/charts/determined/templates/master-config.yaml:41]: Oh, I missed your other comment -- does this run into the same inline issue you mentioned? Don't worry about this.

Comment (aaron276h) [reply helm/charts/determined/templates/master-config.yaml:41]: I got this one to work, key was removing the leading `-` also.

File: helm/charts/determined/templates/master-service.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (aaron276h) [reply helm/charts/determined/templates/master-service.yaml]: Got the inline working now.

**Pull request merged**